model/minsky.cc

```diff
@@ -288,78 +288,87 @@ namespace minsky
+                           {
+                             // if defined, check no other defining variable exists
+                             auto alreadyDefined = canvas.model->findAny
+                               (&GroupItems::items,
+                                [&v](const ItemPtr& j)
+                                {return j.get()!=v && j->variableCast() &&  j->variableCast()->defined();});
+                             if (v->isStock())
+                               {
+                                 if (v->defined() && alreadyDefined && !alreadyDefinedMessageDisplayed)
+                                   {
+                                     message("Integral/Stock variable "+v->name()+" already defined.");
+                                     alreadyDefinedMessageDisplayed=true;
+                                   }
+                                 else if (!v->defined() && !alreadyDefined)
+                                   {
+                                     // need to do this var explicitly, as not currently part of model structure
+                                     if (auto vp=VariablePtr(*i))
+                                       {
+                                         vp.retype(VariableType::flow);
+                                         *i=vp;
+                                         convertVarType(vp->valueId(), VariableType::flow);
+                                       }
+                                   }
+                               }
+                             else if (alreadyDefined)
+                               {
+                                 // delete defining wire from this
+                                 assert(v->ports.size()>1 && !v->ports[1]->wires().empty());
+                                 g->removeWire(*v->ports[1]->wires()[0]);
+                               }
+                           }
+                       return false;
+                     });
+    }                              
 
     canvas.model->addGroup(g); // needed to ensure wires are correctly handled
     auto copyOfItems=g->items;
-
-    for (auto& i: copyOfItems)
-      {		
-         canvas.model->addItem(i);			  
-         canvas.selection.ensureItemInserted(i);
-         assert(!i->ioVar());
-      }
-
+    auto copyOfGroups=g->groups;
+    
+    g->resizeOnContents(); // used in insertGroupFromfile(), so why not here?
+    // ungroup g, putting all its contents on the canvas
+    canvas.model->moveContents(*g); 
+
+    // leave newly ungrouped items in selection
+    for (auto& i: copyOfItems) {
+       canvas.selection.ensureItemInserted(i);
+       // ensure that initial values of pasted parameters are correct. for ticket 1258
+       if (auto v=i->variableCast())
+		 if (v->type()==VariableType::parameter && !existingParms.empty()) 
+		   for (vector<pair<string,string>>::iterator it = existingParms.begin() ; it != existingParms.end(); ++it)
```

Comment:
Why wouldn't you use a std::map for this purpose?

Reply:
Very good question. I was considering it last night, but because I stumbled across the solution so late, I decided to commit it. Changed it now :)